Create participations for exam exercises if they don't exist

TobiasPr opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [x] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) **on test server 3**
- [x] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
<!-- Why is this change required? What problem does it solve? -->
<!-- If it fixes an open issue, please link to the issue here. -->
If a instructor forgets to generate the participations, the participations are generated before starting an exercise on the client

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Prepare an exam (no modeling exercises), but do not generate the participation for the exercises (don't click prepare exercise start)
2. Participating in exam should work as usual
3. If you are offline when navigating to the exercise for the very first time (not reproducible by reloading, the participation for that exercise must be deleted to have a new try) or the preparation (generation of participation or submission does not work) -> Error + retry button is shown - see screenshot
4. Test for programming exercises (special case)

Known issues which will be tackled in follow up PRs:

1. Retrying the generation of participations might fail, if the client for whatever reason don't get the server response, but the server does prepare the exercise (rare edge case)
2. After trying the generation one time offline, the exercise is not drawn immediately when the retry to generate participations works (changing exercise solves that)
3. Programming exercises can be initiated, but the synchronization behavior does not work as expected until the first submission (look at the navigation bar)

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
**Generating Participation**
![image (4)](https://user-images.githubusercontent.com/22955066/86002295-71389c80-ba10-11ea-9c8b-4d1ce70f7baa.png)
**Error occured**
![image](https://user-images.githubusercontent.com/22955066/86002018-07b88e00-ba10-11ea-8816-f457472fae87.png)

TobiasPr commented:
> I've tested this on the test server 2 and it does not work as expected.
> 
> To reproduce:
> 
> 1. Create an exam with two exercise groups (in my example, this was without random order). Add a text exercise to the first exercise group and a quiz exercise to the second exercise group (note: I did not check whether the types of the exercises matter here.)
> 2. Register yourself as a student and generate your student exam without preparing the exercise start.
> 3. Go to the exam and participate.
> 4. The first exercise should be the text exercise. If you switch to the second exercise, it will display the same text exercise again. (Expected: It should display the quiz exercise instead.)

updated it, the navigation bar was resettet when generating a new participation. It should work now for multiple exercises

krusche commented on src/main/webapp/app/exam/participate/exam-participation.component.ts:
```diff
@@ -90,35 +97,17 @@ export class ExamParticipationComponent implements OnInit, OnDestroy {
             this.activeExercise = studentExam.exercises[0];
+            if (this.activeExercise.studentParticipations && this.activeExercise.studentParticipations.length === 0) {
+                // subscribe to execute
+                this.createParticipationForExercise(this.activeExercise).subscribe();
```
There is another edge case. It could be that starting the exercise did not work properly and studentParticipations[0].initializationState is not yet set to INITIALIZED, but to any "previous" enum value.

Then, we also need to call startExercise on the server. The server will automatically determine the right steps to make sure the participation initialization is finished properly.

This is also the case for the course dashboard. If a participation is not fully initialized, Artemis still shows the "start exercise" button. You can also check exercise-utils.ts, however it uses a rather complex calculation which should not be necessary here.

TobiasPr replied:
added a check for that

TobiasPr commented:
> I've encountered another problem while testing this on TS3.
> 
> To reproduce:
> 
> 1. Create an exam with two exercise groups. Add a text exercise to the first and a programming exercise (with the online code editor enabled) to the second.
> 2. Add yourself as a student and generate the exams without preparing them
> 3. Participate in the exam, the text exercise should work as expected.
> 4. Switch to the programming exercise, wait for the participation to be generated, then edit something and submit it.
> 5. Press "Save" in the exam header bar.
> 6. Switch back to the previous exercise.
> 7. Artemis displays the programming exercise. (Expected: Artemis should display the text exercise).
> 
> If you then refresh the page and re-enter, the bug will no longer occur. There are error messages when saving and when switching between the exercises:
> 
> When saving the programming exercise via the header bar:
> 
> ![image](https://user-images.githubusercontent.com/38403547/86116408-ad7e0280-bacd-11ea-93c9-62630b5806a6.png)
> 
> When returning to the first exercise:
> 
> ![image](https://user-images.githubusercontent.com/38403547/86116431-b7076a80-bacd-11ea-9445-8fef8274d67c.png)

added a check for that. Tested on test server 2. After the participation was created I added a file to programming exercise did some changes in .java file, did a build and after that changed exercise which worked for me -> so I think it is fine now

TobiasPr commented:
> I have tested this on TS3 and I was unable to produce any more unwanted behavior that relates to this pull request.
> 
> (Probably unrelated:) I'd like to mention that programming exercises seem to currently break the summary.
> (What I did: "Save" a programming exercise through the header bar, which produced an error in my console. Submitted some changes normally. Waited until the exam was over, the summary opened and the programming exercise's summary was not displayed. Instead, my console was flooded with errors).
> 
> This appears to happen regardless of whether I prepare the exercise start or not.
> 
> When hitting "Save" for a programming exercise via the header bar
> 
> ![image](https://user-images.githubusercontent.com/38403547/86158960-e0de8280-bb09-11ea-80f4-abad3937f2a9.png)
> 
> Once the summary opens:
> 
> ![image](https://user-images.githubusercontent.com/38403547/86159083-0ec3c700-bb0a-11ea-96b6-fe53c275d5ef.png)

I added some checks to make the summary page more robust

tiffca commented:
Tested this very briefly with Tobias, please let me know if I should test further !

jpbernius commented on src/main/webapp/app/exam/participate/exam-participation.component.ts:
```diff
@@ -27,6 +30,8 @@ import { throttleTime } from 'rxjs/operators';
 import * as moment from 'moment';
 import { Moment } from 'moment';
 
+type GenerateParticipationStatus = 'generating' | 'failed' | 'success';
```
:+1:

TobiasPr replied:
not entirely sure if you mean that ironic :)

jpbernius replied:
Nah, I like those 🙂

TobiasPr commented:
> ts, code looks good otherwise. Still need to test this.

there is a issue that not the latest submission cannot be retrieved when you go into online editor. So the synchronization state does not work as expected. I am working on that in a follow up

krusche commented:
Seems that the code still has some problems, so I changed it back to a draft PR.
@TobiasPr as soon as you have time again: please double check that everything (except of really strange edge cases) work correctly

TobiasPr commented:
The issue @pzdr7 mentioned happened, because we introduced new code from develop which assumed, that a submission is always available when participating in an exam. I fixed that now and will test all scenarios again on a testserver. We should try to merge this then soon, because it can happen, that code changes from develop do break this branch again even if there is no merge conflict. This can happen, because we changed a key assumption in the client, which is, that a participation and submission is available at all times for each exam exercise

krusche commented:
> Stated behaviour with Text/quiz exercises loading without explicitely preparing the exercise start works fine for me.
> There are problems with programming exercises and offline mode like already mentioned (for programming exercises, it loads shortly and afterwards shows me the programming exercise while throwing an 404, see screenshot)
> ![Capture1](https://user-images.githubusercontent.com/44401560/86926047-e2bfcb80-c131-11ea-9046-87753ce51e96.JPG)
> 
> All in all, text exercises and quizes worked fine, though.

Please ignore this error message. It's a weird error of the Ace online code editor that we are using. You will see it in all situations when you first use the online editor. There is some problem with a mode-undefined, we never found where it is coming from. Hopefully soon, we will switch to a more modern code editor

TobiasPr commented:
> quiz/text work well - the exercise generates for the student, the submissions are saved and can be graded.
> 
> **programming exercise does not work** -> the student is not able to view the programming exercise at all. receive this error when selecting "retry" button: "Failed to load resource: the server responded with a status of 403 ()"
> 
> modeling exercise does not work - but this is likely out of scope for pr since it was recently added. error: the exercise does generate and the student can view it, but the student's submissions are not saved.
> 
> Steps (test server 3):
> 
> 1. (instructor) create exam
> 2. create exercise groups
> 3. import text/modeling/programming exercise
> 4. add student
> 5. generate individual exams (do not select 'prepare exam')
> 6. take the exam as a student
> 
> offline mode - steps
> (online) select the exam, start the exam
> (offline/turn off internet) try any exercise -> does not work, as expected. Offline mode should really be an extreme edge case in my opinion. Can't view/select any exercises.
> (go back online) exercises are viewable again, answers are saved

tried as a student on test server 2 and the issue did not appear - @tiffca can you double check?
